### PR TITLE
kucoin tier based fees

### PR DESCRIPTION
**Pull request by samgermain**
I copied the structure from the binance futures tiers, using these [tier prices](https://www.kucoin.com/vip/level), but there's a lot of information missing, I feel like a tier structure that looks more like this would be better

```
'tiers': [
    {
        'minTradeVolume': 0,
        'minFuturesTradeVolume': 0,
        'takerFee': 0.001,
        'makerFee': 0.001,
        'minExchangeCurrency': 0,
        'exchangeCurrencyTakerFee': 0.0008,
        'exchangeCurrencyMakerFee': 0.0008,
    },
    {
        'minTradeVolume': 50,
        'minFuturesTradeVolume': 100,
        'takerFee': 0.001,
        'makerFee': 0.0009,
        'minExchangeCurrency': 1000,
        'exchangeCurrencyTakerFee': 0.0008,
        'exchangeCurrencyMakerFee': 0.00072,
    },
    {
        'minTradeVolume': 200,
        'minFuturesTradeVolume': 400,
        'takerFee': 0.0009,
        'makerFee': 0.0007,
        'minExchangeCurrency': 10000,
        'exchangeCurrencyTakerFee': 0.00072,
        'exchangeCurrencyMakerFee': 0.00056,
    },
    {
        'minTradeVolume': 500,
        'minFuturesTradeVolume': 1000,
        'takerFee': 0.0008,
        'makerFee': 0.0005,
        'minExchangeCurrency': 20000,
        'exchangeCurrencyTakerFee': 0.00064,
        'exchangeCurrencyMakerFee': 0.0004,
    },
    {
        'minTradeVolume': 1000,
        'minFuturesTradeVolume': 2000,
        'takerFee': 0.0007,
        'makerFee': 0.0003,
        'minExchangeCurrency': 30000,
        'exchangeCurrencyTakerFee': 0.00056,
        'exchangeCurrencyMakerFee': 0.00024,
    },
    {
        'minTradeVolume': 2000,
        'minFuturesTradeVolume': 3000,
        'takerFee': 0.0007,
        'makerFee': 0,
        'minExchangeCurrency': 40000,
        'exchangeCurrencyTakerFee': 0.00056,
        'exchangeCurrencyMakerFee': 000,
    },
    {
        'minTradeVolume': 4000,
        'minFuturesTradeVolume': 6000,
        'takerFee': 0.0006,
        'makerFee': 0,
        'minExchangeCurrency': 50000,
        'exchangeCurrencyTakerFee': 0.00048,
        'exchangeCurrencyMakerFee': 000,
    },
    {
        'minTradeVolume': 8000,
        'minFuturesTradeVolume': 12000,
        'takerFee': 0.0005,
        'makerFee': 0,
        'minExchangeCurrency': 60000,
        'exchangeCurrencyTakerFee': 0.0004,
        'exchangeCurrencyMakerFee': 000,
    },
    {
        'minTradeVolume': 15000,
        'minFuturesTradeVolume': 20000,
        'takerFee': 0.00045,
        'makerFee': -0.00005,
        'minExchangeCurrency': 70000,
        'exchangeCurrencyTakerFee': 0.00036,
        'exchangeCurrencyMakerFee': -0.00005,
    },
    {
        'minTradeVolume': 25000,
        'minFuturesTradeVolume': 40000,
        'takerFee': 0.0004,
        'makerFee': -0.00005,
        'minExchangeCurrency': 80000,
        'exchangeCurrencyTakerFee': 0.00032,
        'exchangeCurrencyMakerFee': -0.00005,
    },
    {
        'minTradeVolume': 40000,
        'minFuturesTradeVolume': 80000,
        'takerFee': 0.00035,
        'makerFee': -0.00005,
        'minExchangeCurrency': 90000,
        'exchangeCurrencyTakerFee': 0.00028,
        'exchangeCurrencyMakerFee': -0.00005,
    },
    {
        'minTradeVolume': 60000,
        'minFuturesTradeVolume': 120000,
        'takerFee': 0.0003,
        'makerFee': -0.00005,
        'minExchangeCurrency': 100000,
        'exchangeCurrencyTakerFee': 0.00024,
        'exchangeCurrencyMakerFee': -0.00005,
    },
    {
        'minTradeVolume': 80000,
        'minFuturesTradeVolume': 160000,
        'takerFee': 0.00025,
        'makerFee': -0.00005,
        'minExchangeCurrency': 150000,
        'exchangeCurrencyTakerFee': 0.0002,
        'exchangeCurrencyMakerFee': -0.00005,
    },
]
```

### CODE REVIEW

**Comment by kroitor** [comment]
> I feel like a tier structure that looks more like this would be better

The maker-taker fee ladders don't always align in volumes. So we better keep them separate for now.
On a sidenote let's always use this.parseNumber and string-based values as much as we can: https://github.com/ccxt/ccxt/pull/10791/commits/8832c43b687fd3f685e75bc14633ba79d46977b0

File: js/kucoin.js [hunk 1]
```diff
@@ -300,38 +300,38 @@ module.exports = class kucoin extends Exchange {
...
+                            [ this.parseNumber ('500'), this.parseNumber ('0.0005') ],
+                            [ this.parseNumber ('1000'), this.parseNumber ('0.0003') ],
+                            [ this.parseNumber ('2000'), this.parseNumber ('0') ],
+                            [ this.parseNumber ('4000'), this.parseNumber ('0') ],
+                            [ this.parseNumber ('8000'), this.parseNumber ('0') ],
+                            [ this.parseNumber ('15000'), this.parseNumber ('-0.00005') ],
+                            [ this.parseNumber ('25000'), this.parseNumber ('-0.00005') ],
+                            [ this.parseNumber ('40000'), this.parseNumber ('-0.00005') ],
+                            [ this.parseNumber ('60000'), this.parseNumber ('-0.00005') ],
+                            [ this.parseNumber ('80000'), this.parseNumber ('-0.00005') ],
```
Comment (samgermain) [inline js/kucoin.js:334]: Why do we want this instead?

Comment (kroitor) [reply js/kucoin.js:334]: We are going to switch to strings everywhere:

- https://github.com/ccxt/ccxt/issues/522
- https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#adding-numbers-and-concatenating-strings

So, `parseNumber` currently returns native numbers, but will also return strings, depending on the user preference.

**Comment by samgermain** [comment]
> > I feel like a tier structure that looks more like this would be better
> 
> The maker-taker fee ladders don't always align in volumes. So we better keep them separate for now. On a sidenote let's always use this.parseNumber and string-based values as much as we can: [8832c43](https://github.com/ccxt/ccxt/commit/8832c43b687fd3f685e75bc14633ba79d46977b0)

The main problem I'm trying to point out is that there's missing data, you can get to a higher tier by having a min trade volume of ___ OR a min trade volume of ___ and we're not including any information about the exchange currencies

**Comment by kroitor** [comment]
> The main problem I'm trying to point out is that there's missing data, you can get to a higher tier by having a min trade volume of ___ OR a min trade volume of ___ and we're not including any information about the exchange currencies

That's true, but in order to devise a new unified structure we should review the majority of the exchanges that have fee tiers.

**Comment by kroitor** [comment]
> you can get to a higher tier by having a min trade volume of ___ OR a min trade volume of ___

Min 30-day volumes are actually accounted-for in the existing tiers structure.

**Comment by kroitor** [comment]
We can also add an extra level of depth to differentiate spot fees and tiers from futures fees and tiers, for example:

- https://github.com/ccxt/ccxt/blob/master/js/okex3.js#L310
- https://github.com/ccxt/ccxt/blob/master/js/binance.js#L600

File: js/kucoin.js (hunk 1 above)
Comment (samgermain) [reply js/kucoin.js:334]: So every number in CCXT should be `this.parseNumber ('...')`

Comment (kroitor) [reply js/kucoin.js:334]: @samgermain yes, every number in CCXT should eventually become `this.safeNumber (obj, key)` or `this.parseNumber (string)`.

**Pull request merged**